src/unlock/opentx/hasher.rs

```diff
@@ -0,0 +1,716 @@
+use anyhow::anyhow;
+use bitflags::bitflags;
+use bytes::{BufMut, Bytes};
+use ckb_hash::Blake2b;
+use ckb_types::{bytes::BytesMut, core::TransactionView, prelude::*};
+use serde::{Deserialize, Serialize};
+use std::convert::TryFrom;
+
+use enum_repr_derive::{FromEnumToRepr, TryFromReprToEnum};
+
+use super::reader::{OpenTxCellField, OpenTxReader, OpenTxSource};
+use super::OpenTxError;
+
+const ARG1_MASK: u16 = 0xFFF;
+const ARG2_MASK: u16 = 0xFFF;
+
+/// Open transaction signature input command.
+#[derive(
+    Clone,
+    Copy,
+    Serialize,
+    Deserialize,
+    Debug,
+    Hash,
+    Eq,
+    PartialEq,
+    TryFromReprToEnum,
+    FromEnumToRepr,
+)]
+#[repr(u8)]
+pub enum OpentxCommand {
+    /// Hash the full current transaction hash
+    TxHash = 0x00,
+    /// Hash length of input & output cells in current script group
+    GroupInputOutputLen = 0x01,
+    /// Hash part or the whole output cell. arg1 is index of output cell, arg2 is cell mask.
+    IndexOutput = 0x11,
+    /// Hash part or the whole output cell. arg1 is offset of output cell, arg2 is cell mask.
+    OffsetOutput = 0x12,
+    /// Hash part or the whole input cell. arg1 is index of input cell, arg2 is cell mask.
+    IndexInput = 0x13,
+    /// Hash part or the whole input cell. arg1 is offset of input cell, arg2 is cell mask.
+    OffsetInput = 0x14,
+    /// Hash part or the whole cell input structure, arg1 is index of input cell, arg2 is input mask.
+    CellInputIndex = 0x15,
+    /// Hash part or the whole cell input structure, arg1 is offset of input cell, arg2 is input mask.`
+    CellInputOffset = 0x16,
+    /// Concatenate ARG 1 and ARG 2, arg1 is lower 12 bit, arg2 is higher 12 bit.
+    /// The purpose of this command is to add salt for hash.
+    ConcatArg1Arg2 = 0x20,
+    /// Terminate and generate the final blake2b hash
+    End = 0xF0,
+}
+
+impl OpentxCommand {
+    pub fn is_index(&self) -> bool {
+        matches!(
+            self,
+            OpentxCommand::IndexOutput | OpentxCommand::IndexInput | OpentxCommand::CellInputIndex
+        )
+    }
+}
+
+bitflags! {
+    /// The bits control the data to generate from a cell.
+    #[derive(Serialize, Deserialize)]
+    pub struct CellMask: u16 {
+        /// capacity
+        const CAPACITY = 0x1;
+        /// lock.code_hash
+        const LOCK_CODE_HASH = 0x2;
+        /// lock.hash_type
+        const LOCK_HASH_TYPE = 0x4;
+        /// lock.args
+        const LOCK_ARGS = 0x8;
+        /// type.code_hash
+        const TYPE_CODE_HASH = 0x10;
+        /// type.hash_type
+        const TYPE_HASH_TYPE = 0x20;
+        /// type.args
+        const TYPE_ARGS = 0x40;
+        /// Cell data
+        const CELL_DATA = 0x80;
+        /// Lock script hash
+        const TYPE_SCRIPT_HASH = 0x100;
```

Comment:
```suggestion
        const LOCK_SCRIPT_HASH = 0x100;
```

Reply:
Some definitions in the omni lock script code have been modified: https://github.com/nervosnetwork/ckb-production-scripts/commit/e848f4feca47c03800d3822eb519c7f5fd26f191